src/wazuh_db/wdb_global.c

```diff
@@ -1845,42 +1849,47 @@ cJSON* wdb_global_get_backups() {
 }
 
 int wdb_global_restore_backup(wdb_t** wdb, char* snapshot, bool save_pre_restore_state, char* output) {
+    char* backup_to_restore = NULL;
+
+    // If the snapshot is not present, the most recent backup will be used
+    if (snapshot) {
+        backup_to_restore = snapshot;
+    } else {
+        wdb_global_get_most_recent_backup(&backup_to_restore);
+    }
+
     char global_path[OS_SIZE_256] = {0};
-    char global_pre_restore_path[OS_SIZE_256] = {0};
 
     snprintf(global_path, OS_SIZE_256, "%s/%s.db", WDB2_DIR, WDB_GLOB_NAME);
-    snprintf(global_pre_restore_path, OS_SIZE_256, "%s/%s.gz", WDB_BACKUP_FOLDER, WDB_GLOB_PRE_RESTORE_BACKUP_NAME);
 
-    // Preparing DB for pre_restore backup and later removal
-    wdb_leave(*wdb);
-    wdb_close(*wdb, true);
-    *wdb = NULL;
-    if(save_pre_restore_state) {
-        unlink(global_pre_restore_path);
-        if(w_compress_gzfile(global_path, global_pre_restore_path)) {
-            mdebug1("Unable to save pre-restore DB state");
-            snprintf(output, OS_MAXSTR + 1, "err Unable to save pre-restore DB state");
-            return OS_INVALID;
+    cJSON* response = NULL;
+    char* payload = NULL;
+
+    if (save_pre_restore_state) {
+        if (OS_SUCCESS == wdb_global_create_backup(*wdb, output, "-pre_restore")) {
+            wdbc_parse_result(output, &payload);
+            response = cJSON_Parse(payload);
+            mdebug1("Creating pre-restore global DB snapshot %s", cJSON_GetStringValue(cJSON_GetArrayItem(response, 0)));
+            cJSON_Delete(response);
+        } else {
+            mwarn("Creating pre-restore global DB snapshot failed");
         }
     }
 
-    // If the snapshot is not present, the most recent backup will be used
-    char* backup_to_restore = NULL;
-    if(snapshot) {
-        backup_to_restore = snapshot;
-    } else {
-        wdb_global_get_most_recent_backup(&backup_to_restore);
-    }
+    // Preparing DB for restore backup and later removal.
+    wdb_leave(*wdb);
```

Comment:
Maybe, we can move these 3 steps just below the `if (!w_uncompress_gzfile(backup_to_restore_path, global_tmp_path)) {`, to do them only if we are sure nothing else failed.

Reply:
Changed